src/collapse/dtrace.rs

```diff
@@ -0,0 +1,187 @@
+use super::Collapse;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+// Most code should not have more then 10 calls in the java stack.
+const JAVA_VEC_SIZE: usize = 10;
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// include function offset (except leafs)
+    pub includeoffset: bool,
+}
+
+/// A stack collapser for the output of dtrace `ustrace()`.
+///
+/// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
+/// `dtrace::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Keep track of stack string size while we consume a stack
+    stack_str_size: usize,
+
+    /// vecotr for processing java stuff
+    java_vec: Vec<String>,
+
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+
+        // skip header lines -- first empty line marks start of data
+        loop {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                // We reached the end :( this should not happen.
+                warn!("File ended while skipping headers");
+                return Ok(());
+            };
+            if line.trim().is_empty() {
+                break;
+            }
+        }
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Self {
+            stack: VecDeque::default(),
+            occurrences: HashMap::default(),
+            java_vec: Vec::with_capacity(JAVA_VEC_SIZE),
+            opt: opt,
+            stack_str_size: 0,
+        }
+    }
+}
+
+impl Folder {
+    fn uncpp(probe: &str) -> &str {
+        if let Some(scope) = probe.rfind("::") {
+            if let Some(open) = probe[scope + 2..].find(|c| c == '(' || c == '<') {
```

Comment:
Based on the conversation [here](https://github.com/jonhoo/inferno/pull/78#discussion_r264042450), I believe this should be:
```rust
if let Some(scope) = frame.find("::") {
    if let Some(open) = frame[scope + 2..].rfind(|c| c == '(' || c == '<') {
```
Notice the two swaps of `rfind` and `find`.

Reply:
I've added some tests, `find` vs `rfind` makes no difference in them, so both seems fine. I'll change it to `find`.

Reply:
Ah, no, the second one specifically needs to be `rfind` for this to correctly match the regex :)

Reply:
👍

Reply:
That will re-add the <...> to the data, is that what we want? I'm genuenly unsure since I couldn't find any examples on C++ dtrace stack traces.

Reply:
Yeah, let's stick with the way the original Perl regex operates here, which would be (I think) exactly like `find` followed by `rfind`. And then we can refine later in another PR if we come up with compelling reasons why it should be different.

Reply:
okay